r/configure

```diff
@@ -58,191 +100,282 @@ if [ ! "`${R_HOME}/bin/R CMD config CXX17`" ]; then
+  if grep -q "_GLIBCXX_USE_CXX11_ABI=0" "${LIB_DIR}/pkgconfig/arrow.pc"; then
+    PKG_CFLAGS="${PKG_CFLAGS} -D_GLIBCXX_USE_CXX11_ABI=0"
+  fi
+  PKG_DIRS="-L${LIB_DIR}"
+  if [ "${OPENSSL_ROOT_DIR}" != "" ]; then
+    PKG_DIRS="${PKG_DIRS} -L${OPENSSL_ROOT_DIR}/lib"
+  fi
+  PKG_LIBS="-larrow"
+  if [ -n "$(find "$LIB_DIR" -name 'libarrow_bundled_dependencies.*')" ]; then
+    PKG_LIBS="$PKG_LIBS -larrow_bundled_dependencies"
+  fi
+
+  # If on Raspberry Pi, need to manually link against latomic
+  # See: https://gcc.gnu.org/bugzilla/show_bug.cgi?id=81358 for similar example
+  # pkg-config will handle this for us automatically, see ARROW-6312
+  if grep raspbian /etc/os-release >/dev/null 2>&1; then
+    PKG_CFLAGS="$PKG_CFLAGS -DARROW_CXXFLAGS=-latomic"
+    PKG_LIBS="-latomic $PKG_LIBS"
+  fi
+}
+
+add_feature_flags () {
+  # Now we need to check what features it was built with and enable
+  # the corresponding feature flags in the R bindings (-DARROW_R_WITH_stuff).
+  # We do this by inspecting ArrowOptions.cmake, which the libarrow build
+  # generates.
+  ARROW_OPTS_CMAKE="$LIB_DIR/cmake/Arrow/ArrowOptions.cmake"
+  if [ ! -f "${ARROW_OPTS_CMAKE}" ]; then
+    echo "*** $ARROW_OPTS_CMAKE not found; some features will not be enabled"
+  else
+    if arrow_built_with ARROW_PARQUET; then
+      PKG_CFLAGS="$PKG_CFLAGS -DARROW_R_WITH_PARQUET"
+      PKG_LIBS="-lparquet $PKG_LIBS"
+      # NOTE: parquet is assumed to have the same -L flag as arrow
+      # so there is no need to add its location to PKG_DIRS
+    fi
+    if arrow_built_with ARROW_DATASET; then
+      PKG_CFLAGS="$PKG_CFLAGS -DARROW_R_WITH_DATASET"
+      PKG_LIBS="-larrow_dataset $PKG_LIBS"
+      # NOTE: arrow-dataset is assumed to have the same -L flag as arrow
+      # so there is no need to add its location to PKG_DIRS
+    fi
+    if arrow_built_with ARROW_ACERO; then
+      PKG_CFLAGS="$PKG_CFLAGS -DARROW_R_WITH_ACERO"
+      PKG_LIBS="-larrow_acero $PKG_LIBS"
+      # NOTE: arrow-acero is assumed to have the same -L flag as arrow
+      # so there is no need to add its location to PKG_DIRS
+    fi
+    if arrow_built_with ARROW_SUBSTRAIT; then
+      PKG_CFLAGS="$PKG_CFLAGS -DARROW_R_WITH_SUBSTRAIT"
+      PKG_LIBS="-larrow_substrait $PKG_LIBS"
+      # NOTE: arrow-substrait is assumed to have the same -L flag as arrow
+      # so there is no need to add its location to PKG_DIRS
+    fi
+    if arrow_built_with ARROW_JSON; then
+      PKG_CFLAGS="$PKG_CFLAGS -DARROW_R_WITH_JSON"
+    fi
+    if arrow_built_with ARROW_S3; then
+      PKG_CFLAGS="$PKG_CFLAGS -DARROW_R_WITH_S3"
+      PKG_LIBS="$PKG_LIBS $S3_LIBS"
+    fi
+    if arrow_built_with ARROW_GCS; then
+      PKG_CFLAGS="$PKG_CFLAGS -DARROW_R_WITH_GCS"
+      PKG_LIBS="$PKG_LIBS $GCS_LIBS"
+    fi
+  fi
+}
+
+arrow_built_with() {
+  # Function to check cmake options for features
+  grep -i 'set('"$1"' "ON")' $ARROW_OPTS_CMAKE >/dev/null 2>&1
+}
+
+##############
+# Main logic #
+##############
```

Comment:
This is a little pedantic, but I wonder if it would be more straightforward to have this up top and the functions down below (kinda like you define `arrow_built_with` after the function that uses it)? 

I'm not absolutely tied to it, but having the operative thing up top and supporting things below it is a nice way to read things

Reply:
Yeah I agree that would be nice.